javaagent-muzzle-generation/build.gradle.kts

```diff
@@ -0,0 +1,22 @@
+plugins {
+  `kotlin-dsl`
+
+  id("otel.java-conventions")
+  id("otel.publish-conventions")
+  id("com.gradle.plugin-publish")
+}
+
+group = "io.opentelemetry.instrumentation.gradle"
+
+dependencies {
+  implementation("com.google.guava:guava")
```

Comment:
Sorry had a pending comment but forgot to hit submit. On the topic of a `codegen` configuration

I think we may still want the configuration

1) To add less to Gradle classpath. If we add too much it can cause dependency hell, a bit worried about Guava
2) To be able to substitute in the project dependencies in our project, not sure we'd have any way of doing it for a plugin transitive dependency which would be resolved way too early for a substitution to kick in I think.

I guess 1 could be resolved with shading anyways though. But 2) seems important and a requirement to allow using an older version of the plugin even as we work on the javaagent-extension-api.

Reply:
At the same time self-contained plugin is good for external users. They don't need to know about "codegen" dependency and that something has to be added to it.

Reply:
Yeah they wouldn't need to with the configuration either since we configure it with defaults here. Before this PR our repo is substituting to use our project dependencies but it's not required to care otherwise.

Reply:
I am afraid I don't quite understand what do you suggest? Do you mean to create and use `codegen` configuration already here, right in the module? Or how?

Reply:
I propose using the same pattern we have

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/buildSrc/src/main/kotlin/io.opentelemetry.instrumentation.javaagent-testing.gradle.kts#L33

This as all the io.openteletry.* ones are meant to be published for downstream. Though most will get this one transitively via io.opentelemetry.instrumentation.javaagent-instrumentation.

The configuration is set here when the plugin is included so most users won't need to worry about it.

Reply:
Sorry by here I meant there :P I mean the config is set in the javaagent-testing plugin

Reply:
Also makes me realize this project shouldn't be called muzzle or codegen or anything - let's call it `gradle-plugins`. We have a few plugins to publish.

Reply:
> Also makes me realize this project shouldn't be called muzzle or codegen or anything - let's call it `gradle-plugins`. We have a few plugins to publish.

Which ones? I am fine with publishing muzzle generation and checking as one plugin. But I am not sure we should bundle all our plugins into one artifact?

Reply:
> I propose using the same pattern we have
> 
> https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/buildSrc/src/main/kotlin/io.opentelemetry.instrumentation.javaagent-testing.gradle.kts#L33
> 
> This as all the io.openteletry.* ones are meant to be published for downstream. Though most will get this one transitively via io.opentelemetry.instrumentation.javaagent-instrumentation.
> 
> The configuration is set here when the plugin is included so most users won't need to worry about it.

Do I understand you correctly?

- Published `io.opentelemetry.instrumentation.javaagent-codegen` plugin does not have any declared runtime dependencies
- Actual dependencies, such as guava and `javaagent-extension-api` are added by `io.opentelemetry.instrumentation.javaagent-instrumentation` plugin
- End users are supposed to use `io.opentelemetry.instrumentation.javaagent-instrumentation` plugin when it gets published

Reply:
Yeah that understanding sounds correct to me.

>  But I am not sure we should bundle all our plugins into one artifact?

I don't think there's any real downside - anyways we'll publish to Gradle Plugin Portal so users just add the plugin to their project and don't care how the artifact is structured. So it allows us to keep all plugin publishing type of logic together.

Reply:
First, we don't actually HAVE to publish to Gradle Plugin Portal. Plugins can be consumed just from Maven Central as well.

Second, ok, let's give it a try...